tools/get-python-bin.sh

```diff
@@ -0,0 +1,45 @@
+#!/usr/bin/env bash
+
+# Synopsis:
+# - get-python-bin.sh python
+# - get-python-bin.sh pip
+#
+# This script selects the most suitable `python` / `pip` binary available
+# for building docs that's installed in this system.
+#
+# If no usable version of Python is available, this script will exit with a
+# non-zero code.
+
+set -e
+
+case "$1" in
+    "python" | "pip")
+        BIN_PREFIX="$1"
+        ;;
+    *)
+        echo "Incorrect usage" >&2
+        exit 1
+        ;;
+esac
+
+# if `python3` is >= 3.8 and `pip3` is available, use that
+if echo -e "import sys\nif sys.version_info < (3,8):\n exit(1)" | python3 && \
+which pip3 1>/dev/null 2>&1; then
+    echo "${BIN_PREFIX}3"
+    exit 0
+fi
+
+# check `python3.N` from newest to oldest
+CANDIDATE_SUFFIXES=("3.11" "3.10" "3.9" "3.8")
```

Comment:
This list would have a maintenance overhead from having to update to accept new python versions every time a new version of Python 3 is released. Instead, you should probably do something like `python3 --version | grep "^Python 3\.[0-7][$|\.]"` to check if the version of Python matches a known incompatible version and perform the error actions if it does.

Reply:
I understand what you're talking about - in fact I initially wanted to do something like this. This is what I would write if we are to go down this route:

```bash
# compgen is bash-specific, but we asked for bash in shebang so it's fine
CANDIDATE_MINOR_SUFFIXES=$(compgen -c | grep -P '^python3\.[0-9]+$' | sed 's/python3\.//' | awk 'int($NF) >= 8' | sort -nr)
```

`compgen` gives all executables in `$PATH`; `grep` filters for `python3.*`; `sed` strips prefix; `awk` sets minimum version; `sort` orders the versions.

---

The reason I eventually decided against this is for the simple reason that Python is not exactly known for strong future compatibility guarantees. While unlikely, there is a non-insignificant chance that Python 3.14 or something comes along and breaks standard library (for an example of this see https://github.com/bastibe/SoundCard/issues/140).

So in the end I simply verified that all of 3.8, 3.9, 3.10, and 3.11 work, then hard-coded them here. It's just not worth it to try to anticipate what will happen in the future. When 3.12 / 3.13 / 3.14 eventually comes along, it would just be a trivial edit anyways.

Reply:
On second thought, maybe the best way to handle this is to automatically use the newest (and possibly future) versions, but also allow overriding in Makefile. This way in case a future python version breaks something, packagers can always set a temporary override without having to wait for an upstream fix.

I'll go ahead and make these changes for now.